services/org/org.go

```diff
@@ -15,12 +15,44 @@ import (
 	user_model "code.gitea.io/gitea/models/user"
 	"code.gitea.io/gitea/modules/storage"
 	"code.gitea.io/gitea/modules/util"
-	user_service "code.gitea.io/gitea/services/user"
+	repo_service "code.gitea.io/gitea/services/repository"
 )
 
 // DeleteOrganization completely and permanently deletes everything of organization.
-func DeleteOrganization(org *org_model.Organization) error {
-	ctx, commiter, err := db.TxContext(db.DefaultContext)
+func DeleteOrganization(ctx context.Context, org *org_model.Organization, purge bool) error {
+	if purge {
+		// Delete all repos belonging to this organisation
+		// Now this is not within a transaction because there are internal transactions within the DeleteRepository
+		// BUT: the db will still be consistent even if a number of repos have already been deleted.
+		// And in fact we want to capture any repositories that are being created in other transactions in the meantime
+		//
+		// An alternative option here would be write a DeleteAllRepositoriesForUserID function which would delete all of the repos
+		// but such a function would likely get out of date
+		for {
+			repos, _, err := repo_model.GetUserRepositories(&repo_model.SearchRepoOptions{
+				ListOptions: db.ListOptions{
+					PageSize: repo_model.RepositoryListDefaultPageSize,
+					Page:     1,
+				},
+				Private: true,
+				OwnerID: org.ID,
+				Actor:   org.AsUser(),
+			})
+			if err != nil {
+				return fmt.Errorf("GetUserRepositories: %w", err)
+			}
+			if len(repos) == 0 {
+				break
+			}
+			for _, repo := range repos {
+				if err := repo_service.DeleteRepositoryDirectly(ctx, org.AsUser(), org.ID, repo.ID); err != nil {
+					return fmt.Errorf("unable to delete repository %s for %s[%d]. Error: %w", repo.Name, org.Name, org.ID, err)
+				}
+			}
+		}
+	}
+
+	ctx, commiter, err := db.TxContext(ctx)
```

Comment:
I think `DeleteRepositoryDirectly` should be also under the transaction. Otherwise, if there will be a dirty situation that all repositories deleted but organization related information hasn't been deleted.

Reply:
`DeleteRepositoryDirectly` uses a transaction, so we would have a transaction in a transaction. Not sure if that's causing problems.

Reply:
@JakobDev nested transactions in `sql` are normal thing.

Reply:
I have now tried to call `DeleteRepositoryDirectly` in the transaction, but it resulted in a deadlock in SQLite

Reply:
@JakobDev anything out of ordinary in log?

Reply:
No, it just hangs forever. If you are interested, you can easily try it out yourself.

Reply:
> No, it just hangs forever. If you are interested, you can easily try it out yourself.

`ctx, commiter, err := db.TxContext(ctx)` will check whether `ctx` contains a transaction, if it contains, it will reuse the transaction, otherwise it will create one.

Reply:
It is used in may places, but this particular place causes a deadlock, so we can't use it here.